chalk-ir/src/lib.rs

```diff
@@ -271,6 +265,244 @@ impl<I: Interner> TyData<I> {
+    /// at the innermost binder. Otherwise, returns `None`.
+    pub fn index_if_bound_at(self, debruijn: DebruijnIndex) -> Option<usize> {
+        if self.debruijn == debruijn {
+            Some(self.index)
+        } else {
+            None
+        }
+    }
+}
+
+/// References the binder at the given depth. The index is a [de
+/// Bruijn index], so it counts back through the in-scope binders,
+/// with 0 being the innermost binder. This is used in impls and
+/// the like. For example, if we had a rule like `for<T> { (T:
+/// Clone) :- (T: Copy) }`, then `T` would be represented as a
+/// `BoundVar(0)` (as the `for` is the innermost binder).
+///
+/// [de Bruijn index]: https://en.wikipedia.org/wiki/De_Bruijn_index
+#[derive(Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct DebruijnIndex {
+    depth: u32,
+}
+
+impl From<u32> for DebruijnIndex {
+    fn from(depth: u32) -> Self {
+        Self::from_u32(depth)
+    }
+}
+
+impl From<usize> for DebruijnIndex {
+    fn from(depth: usize) -> Self {
+        assert!(depth < (std::u32::MAX as usize));
+        Self::from_u32(depth as u32)
+    }
+}
+
+impl DebruijnIndex {
+    pub const INNERMOST: DebruijnIndex = DebruijnIndex { depth: 0 };
+
+    pub fn from_u32(depth: u32) -> Self {
+        DebruijnIndex { depth }
+    }
+
+    pub fn as_u32(self) -> u32 {
+        self.depth
+    }
+
+    pub fn as_usize(self) -> usize {
+        self.depth as usize
+    }
+
+    /// True if this represents a binder that is "within" the
+    /// `binders` innermost binders.
+    ///
+    /// # Example
+    ///
+    /// Imagine you have the following binders in scope
+    ///
+    /// ```ignore
+    /// forall<a> forall<b> forall<c>
+    /// ```
+    ///
+    /// then the Debruijn index for `c` would be `0`, the index for
+    /// `b` would be 1, and so on. Now consider the following calls:
+    ///
+    /// * `c.within(0) = false` -- nothing is within the 0 innermost binders
+    /// * `c.within(1) = true` -- but c is within the innermost binder
+    /// * `b.within(1) = false` -- b is not
+    /// * `b.within(2) = true` -- b is within the 2 innermost binders
+    /// * `b.within(3) = true` -- b is also within the 3 innermost binders
+    pub fn within(self, binders: usize) -> bool {
+        self.as_usize() < binders
+    }
+
+    /// True if the binder identified by this index is within the
+    /// binder identified by the index `binder`.
+    ///
+    /// # Example
+    ///
+    /// Imagine you have the following binders in scope
+    ///
+    /// ```ignore
+    /// forall<a> forall<b> forall<c>
+    /// ```
+    ///
+    /// then the Debruijn index for `c` would be `0`, the index for
+    /// `b` would be 1, and so on. Now consider the following calls:
+    ///
+    /// * `c.within(a) = true`
+    /// * `b.within(a) = true`
+    /// * `a.within(a) = false`
+    /// * `a.within(c) = false`
+    pub fn within_binder(self, binder: DebruijnIndex) -> bool {
```

Comment:
I think what a lot of these comments are getting at: My inclination for debruijn index *especially* (but really for a lot of chalk, though it's not always achieved), is that it's worth trying to keep the client code as free of "random conversions" and other noise as possible. I guess I feel like the code tends to be subtle, and seeing things like `foo.within_binder(bar)` instead of `foo.within(bar.as_usize())` that much clearer to me.

That said, I think I was intending to go back and try to make it so that **whenever** we are counting binders, we do it with a debruijn index (in other words, try to remove `within`, instead of `within_binder`). This might also mean we can get rid of more of the u32/usize methods.

Reply:
Okay, so all uses of `within` are in this impl and `BoundVar`. Let's just rename this to `within`; then rename `within` to `within_usize` (or something to that effect) and make it `pub(crate)`.

Reply:
Pushed an update here ...